cardano-node/src/Cardano/Tracing/Metrics.hs

```diff
@@ -87,3 +99,60 @@ instance All HasKESMetricsData xs => HasKESMetricsData (HardForkBlock xs) where
              => WrapForgeStateInfo blk
              -> K KESMetricsData blk
       getOne = K . getKESMetricsData (Proxy @blk) . unwrapForgeStateInfo
+
+-- | This structure stores counters of blockchain-related events,
+--   per individual forge thread.
+--   These counters are driven by traces.
+data ForgingStats
+  = ForgingStats
+  { fsTxsProcessedNum               :: !(IORef Word64)
+  , fsState                         :: !(IORef (Map ThreadId ForgeThreadStats))
+  }
+
+data ForgeThreadStats = ForgeThreadStats
+  { ftsFirstSlot                    :: !SlotNo
+  , ftsLeadershipChecksInitiated    :: !Word64
+  , ftsNodeCannotForgeNum           :: !Word64
+  , ftsNodeIsLeaderNum              :: !Word64
+  , ftsNodeNotLeaderNum             :: !Word64
+  , ftsBlocksForgedNum              :: !Word64
+  , ftsBlocksForgedInvalidNum       :: !Word64
+  , ftsSlotsMissedNum               :: !Word64
+  }
+
+mkForgingStats :: IO ForgingStats
+mkForgingStats =
+  ForgingStats
+    <$> newIORef 0
+    <*> newIORef mempty
+
+mapForgingStatsTxsProcessed ::
+     ForgingStats
+  -> (Int -> Int)
+  -> IO Int
+mapForgingStatsTxsProcessed fs f =
+  atomicModifyIORef' (fsTxsProcessedNum fs)
+    (fromIntegral &&& id . f . fromIntegral)
```

Comment:
Code golfing, eh? Cute, but maybe not the most readable code?

Also: since you're converting the `Word64` to an `Int` and back here, there's not much point in using a `Word64` in the first place, as you're effectively limiting the range to `Int` here.

So if the range of `Word64` is truly needed, you shouldn't convert it to an `Int` (you could let it be incremented by an `Int` if that's convenient, but then rename it, as that's not `map-`). If you don't need the range, then use `Int` in `ForgeThreadStats`.

Reply:
Will change ADT to `Int`, then -- this was sort-of asking for, but I didn't feel too comfortable with the invasivity -- that'd increase the footprint of the change.